pkg/serviceimport/map.go

```diff
@@ -35,22 +37,21 @@ type DNSRecord struct {
 type clusterInfo struct {
 	record *DNSRecord
 	name   string
-	weight uint64
+	weight float64
 }
 
 type serviceInfo struct {
-	key           string
-	records       map[string]*DNSRecord
-	clustersQueue []clusterInfo
-	rrCount       uint64
-	isHeadless    bool
+	key        string
+	records    map[string]*clusterInfo
+	balancer   *loadbalance.SmoothWeightedRR
+	isHeadless bool
 }
 
-func (si *serviceInfo) buildClusterInfoQueue() {
-	si.clustersQueue = make([]clusterInfo, 0)
-	for cluster, record := range si.records {
-		c := clusterInfo{name: cluster, record: record, weight: 0}
-		si.clustersQueue = append(si.clustersQueue, c)
+func (si *serviceInfo) resetLoadBalancing() {
+	si.balancer.RemoveAll()
+
+	for _, info := range si.records {
+		_ = si.balancer.Add(info, info.weight)
```

Comment:
I suggest to add `info.name` instead of the whole `info` object as it's the `name` that is unique (it's the key for the `records` map). This would also be more efficient than hashing the entire struct.